// minify-images/lib/index.js
'use strict';

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

require('babel-polyfill');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _imagemin = require('imagemin');

var _imagemin2 = _interopRequireDefault(_imagemin);

var _imageminMozjpeg = require('imagemin-mozjpeg');

var _imageminMozjpeg2 = _interopRequireDefault(_imageminMozjpeg);

var _imageminPngquant = require('imagemin-pngquant');

var _imageminPngquant2 = _interopRequireDefault(_imageminPngquant);

var _imageminSvgo = require('imagemin-svgo');

var _imageminSvgo2 = _interopRequireDefault(_imageminSvgo);

var _isPlainObject = require('is-plain-object');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

var _isGlob = require('is-glob');

var _isGlob2 = _interopRequireDefault(_isGlob);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULTS = {
  base: process.cwd(),
  src: './src/**/*.{jpg|png|svg|jpeg}',
  dest: './dist',
  jpg: {
    quality: '90'
  },
  png: {
    quality: '65-80',
    speed: 4
  },
  svg: {
    plugins: [{
      removeTitle: true,
      removeDesc: true,
      removeXMLNS: true
    }]
  }
};

var getPath = function getPath(base, filepath) {
  if (_path2.default.isAbsolute(filepath)) {
    return filepath;
  }
  return _path2.default.resolve(base, filepath);
};

var getPaths = function getPaths(object) {
  var settings = Object.assign({}, {
    base: DEFAULTS.base,
    src: DEFAULTS.src,
    dest: DEFAULTS.dest
  }, object);
  var src = getPath(settings.base, settings.src);
  var dest = getPath(settings.base, settings.dest);

  if (!(0, _isGlob2.default)(src) && !_fs2.default.existsSync(src)) {
    throw new ReferenceError('Provided source (' + src + ') does not exist.');
  }

  if ((0, _isGlob2.default)(dest) || _path2.default.extname(dest)) {
    throw new ReferenceError('Your destination path cannot be a file or glob pattern.');
  }

  if (!_path2.default.extname(src)) {
    src = _path2.default.join(src, '*');
  }

  return [src, dest];
};

var compress = function compress() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if (!(0, _isPlainObject2.default)(options)) {
    throw new TypeError('Given options are not an object.');
  }

  var settings = Object.assign({}, DEFAULTS, options);

  var _getPaths = getPaths(settings),
      _getPaths2 = _slicedToArray(_getPaths, 2),
      src = _getPaths2[0],
      dest = _getPaths2[1];

  return new Promise(function (resolve, reject) {
    (0, _imagemin2.default)([src], dest, {
      plugins: [(0, _imageminMozjpeg2.default)(settings.jpg), (0, _imageminPngquant2.default)(settings.png), (0, _imageminSvgo2.default)(settings.svg)]
    }).then(function (files) {
      resolve(files.map(function (file) {
        return file.path;
      }));
    }).catch(function (error) {
      reject(error);
    });
  }).catch(function (error) {
    throw error;
  });
};

module.exports.getPaths = getPaths;
module.exports.compress = compress;
